electrolytes/solvation_shell_extract.py

```diff
@@ -0,0 +1,266 @@
+from solvation_shell_utils import (
+    expand_shell,
+    renumber_molecules_to_match,
+    filter_by_rmsd,
+    filter_shells_with_solute_atoms,
+)
+from utils import validate_metadata_file
+from schrodinger.comparison import are_conformers
+from schrodinger.application.jaguar.utils import group_with_comparison
+from schrodinger.structutils import rmsd
+
+
+def extract_solvation_shells(
+    input_dir: str,
+    save_dir: str,
+    system_name: str,
+    solute_radii: List[float],
+    solvent_radii: List[float],
+    top_n: int,
+):
+    """
+    Given a MD trajectory in a PDB file, perform a solvation analysis
+    on the specified solute to extract the first solvation shell.
+
+    Args:
+        input_dir: Path to 1) the PDB file containing the MD trajectory (system_output.pdb) and 2) a metadata file (system_metadata.json)
+        save_dir: Directory in which to save extracted solvation shells.
+        system_name: Name of the system - used for naming the save directory.
+        solute_radii: List of shell radii to extract around solutes.
+        solvent_radii: List of shell radii to extract around solvents.
+        top_n: Number of snapshots to extract per topology.
+    """
+
+    # Read a structure and metadata file
+    # TODO: add charges
+    logging.info("Reading structure and metadata files")
+
+    # Read metadata
+    with open(os.path.join(input_dir, "metadata_system.json")) as f:
+        metadata = json.load(f)
+
+    validate_metadata_file(metadata)
+
+    partial_charges = np.array(metadata["partial_charges"])
+
+    solutes = {
+        species: res
+        for res, species, type in zip(
+            metadata["residue"], metadata["species"], metadata["solute_or_solvent"]
+        )
+        if type == "solute"
+    }  # {species: residue name} mapping
+    solute_resnames = list(solutes.values())
+
+    solvents = {
+        species: res
+        for res, species, type in zip(
+            metadata["residue"], metadata["species"], metadata["solute_or_solvent"]
+        )
+        if type == "solvent"
+    }  # {species: residue name} mapping
+    solvent_resnames = list(solvents.values())
+
+    # Read a structure
+    structures = StructureReader(os.path.join(input_dir, "system_output.pdb"))
+
+    # For each solute: extract shells around the solute of some heuristic radii and bin by composition/graph hash
+    # Choose the N most diverse in each bin
+
+    for species, residue in solutes.items():
+        logging.info(f"Extracting solvation shells around {species}")
+        for radius in solute_radii:
+            logging.info(f"Radius = {radius} A")
+            expanded_shells = []
+            for i, st in tqdm(enumerate(structures)):  # loop over timesteps
+                # assign partial charges to atoms
+                for at, charge in zip(st.atom, partial_charges):
+                    at.partial_charge = charge
+
+                if i > 10:  # TODO: fix this
+                    break
```

Comment:
Why aren't you going higher than this? What's broken?

Reply:
This is because the workflow isn't very fast and I wanted to facilitate fast debugging. Will need to profile and find ways to speed up.